SpaceEngeneersQOL/DrillManagement/Trivella.cs

```diff
@@ -18,167 +18,253 @@
+      } else {
+        TurnOff(node as IMyTerminalBlock);
+      }
+    }
+    private void TurnOnObject(object node) {
+      if (node is IMyBlockGroup) {
+        TurnOnGroup(node as IMyBlockGroup);
+      } else {
+        TurnOn(node as IMyTerminalBlock);
+      }
+    }
 
-                return true;
-            }
+    private void IncreaseMaxDistanceAndMove(IMyPistonBase piston)
+    {
+      piston.ApplyAction("IncreaseUpperLimit");
+      piston.ApplyAction("Extend");
+    }
 
-            return false;
-        }
+    private void LogPistonPosition(IMyTextSurface lcd, IMyPistonBase current, List<IMyPistonBase> pistons)
+    {
+      pistons.ForEach((piston) =>
+      {
+        var currentFlag = current.EntityId == piston.EntityId ? "> " : "";
+        lcd.WriteText($"{currentFlag}{piston.CustomName} at {piston.CurrentPosition} / {piston.MaxLimit} | {piston.HighestPosition}\n", true);
+      });
+    }
 
-        private void IncreaseMaxDistanceAndMove(IMyPistonBase piston)
+    private bool assertConfig(IMyTextSurface lcd,object pistons,object drills,object rotor,object light) {
+      if (!CheckExistence(pistons, lcd, "Pistons Group")) return false;
+      if (!CheckExistence(drills, lcd, "Drills Group")) return false;
+      if (!CheckExistence(rotor, lcd, "Rotor")) return false;
+      if (!CheckExistence(light, lcd, "Light")) return false;
+      return true;
+    }
+    public void Main(string argument, UpdateType updateSource)
+    {
+      var pistons = GridTerminalSystem.GetBlockGroupWithName(PistonGroupName);
+      var drills = GridTerminalSystem.GetBlockGroupWithName(DrillsGroupName);
+      var rotor = GridTerminalSystem.GetBlockWithName(DrillRotorName) as IMyMotorStator;
+      object light = GridTerminalSystem.GetBlockWithName(DrillDebugLightName);
+      if (light == null) { light = GridTerminalSystem.GetBlockGroupWithName(DrillDebugLightName); }
+      var lcd = GridTerminalSystem.GetBlockWithName(DrillLcdPanelName) as IMyTextSurface ?? Me.GetSurface(0);
+
+      if (!assertConfig(lcd, pistons, drills, rotor, light)) return;
+
+      TurnOffObject(light);
+
+      var pistonsBlockList = new List<IMyPistonBase> { };
+      var drillsBlockList = new List<IMyTerminalBlock> { };
+
+      pistons.GetBlocksOfType(pistonsBlockList);
+      drills.GetBlocksOfType(drillsBlockList);
+
+      Echo("Pistons Found: " + pistonsBlockList.Count);
+      Echo("Drill Found: " + drillsBlockList.Count);
+
+      // select current piston (ordered by the number in the name)
+      var sortedPistons = pistonsBlockList
+          .Where((pistonBlock) => pistonBlock.MaxLimit < pistonBlock.HighestPosition)
+          .OrderBy((pistonBlock) => getPistonIndexFromName(pistonBlock.CustomName))
+          .ToList();
+      Echo("Not Extended pistons: " + sortedPistons.Count);
+      if (sortedPistons.Count == 0)
+      {
+        Echo("All pistons are fully extended! Exiting.");
+        TurnOnObject(light);
+        return;
+      }

```

Comment:
Questo l'ho messo perché altrimenti mi tornano 0 elementi nella lista quando tutti gli elementi sono estesi al massimo ed andava in errore, così facendo però non si ferma tutto... drovemmo trovare il modo di migliorare questo

Reply:
rimane il bug che non viene aggiornato il log della posizione dei pistoni, quindi l'ultimo rimane a 9.5 / 10